modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/js/components/FileEntryPicker.tsx

```diff
@@ -0,0 +1,351 @@
+				message: Liferay.Language.get(
+					'the-format-is-not-valid-please-upload-a-valid-frontend-token-definition-json-file'
+				),
+			});
+
+			setFrontendTokenDefinitionRefValue(MALFORMED_JSON);
+
+			return;
+		}
+
+		setIsValidatingJSON(true);
+
+		if (!alertRole) {
+			setAlertRole('alert');
+		}
+
+		setFeedback({
+			displayType: 'info',
+			message: Liferay.Language.get(
+				'the-frontend-token-definition-json-file-is-being-uploaded-and-validated'
+			),
+		});
+
+		readInputFile(fileName, target);
+	};
+
+	function readInputFile(fileName: string, input: HTMLInputElement) {
+		if (input.files === null) {
+			return;
+		}
+
+		let file = input.files[0];
+
+		const fileReader = new FileReader();
+
+		fileReader.onload = async function (event) {
+			let frontendTokenDefinitionString = event.target?.result as string;
+
+			if (frontendTokenDefinitionString === '') {
+				file = new File(['{}'], fileName, {
+					type: 'application/json',
+				});
+				frontendTokenDefinitionString = '{}';
+			}
+
+			if (frontendTokenDefinitionRef.current) {
+				frontendTokenDefinitionRef.current.value = frontendTokenDefinitionString;
+			}
+
+			await handleUpload(file);
+
+			setIsValidatingJSON(false);
+		};
+
+		fileReader.readAsText(file);
+	}
+
+	const handlePost = async (body: any, endpoint: string) => {
+		try {
+			const response = await Liferay.Util.fetch(
+				'/o/com-liferay-frontend-token-definition-impl' + endpoint,
+				{
+					body,
+					method: 'POST',
+				}
+			);
+
+			const data = await response.json();
+
+			let displayType: DisplayType = 'success';
+
+			if (!response.ok) {
+				displayType = 'danger';
+
+				setFrontendTokenDefinitionRefValue(MALFORMED_JSON);
+			}
+
+			setFeedback({
+				displayType,
+				message: data.message,
+			});
+		}
+		catch (error) {
+			clearSelection();
+
+			setFeedback({
+				displayType: 'danger',
+				message: Liferay.Language.get(
+					'an-error-has-occurred-while-connecting-to-our-servers'
+				),
+			});
+		}
+	};
+
+	async function handleUpload(
+		file: File,
+		focusSelectFileButton: boolean = true
+	) {
+		const formData = new FormData();
+
+		formData.append('file', file);
+
+		await handlePost(formData, '/validate-file');
+
+		if (focusSelectFileButton) {
+			selectFileButtonRef.current?.focus();
+		}
+	}
+
+	const getClientExtensionFrontendTokenDefinitionInfo = () => {
+		handlePost(
+			Liferay.Util.objectToFormData({
+				companyId,
+				externalReferenceCode,
+			}),
+			'/tokens-info'
+		);
+	};
+
+	const updateFeedbackAfterSubmissionError = (
+		frontendTokenDefinition: string
+	) => {
+		handleUpload(
+			new File([frontendTokenDefinition], fileName, {
+				type: 'application/json',
+			}),
+			false
+		);
+	};
+
+	useEffect(() => {
+		if (!frontendTokenDefinitionRef.current) {
+			return;
+		}
+
+		frontendTokenDefinitionRef.current.value = frontendTokenDefinition;
+
+		if (frontendTokenDefinition && externalReferenceCode) {
+			getClientExtensionFrontendTokenDefinitionInfo();
```

Comment:
Since the dependencies for this useEffect is an empty array, it will only run on the first load of the component. As such, only `getClientExtensionFrontendTokenDefinitionInfo` makes sense to me to be in here. Is any of the other parts needed here?

Reply:
When submit is performed and there is an error, the JSP is 'refreshed', the component is imported again and then rendered, it losts the previous state. The idea is to keep the last file and feedback message that was appearing before the user clicking submit.

If there is an `externalReferenceCode`, that means the CX already exists and the user is just visualizing it. If only the  `frontendTokenDefinition` exists, it means there was an error on the form submit, but JSP stores the value and resend it to the component.

Reply:
This could be solved by simply preventing the form submit (disabling the submit button). But JSP forms doesn't usually prevent submit since there is no real time feedback. Do you think disabling the submit button would be better?